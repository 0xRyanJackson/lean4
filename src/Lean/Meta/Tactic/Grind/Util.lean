/-
Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Lean.Meta.AbstractNestedProofs
import Lean.Meta.Tactic.Util

namespace Lean.Meta.Grind

def _root_.Lean.MVarId.transformTarget (mvarId : MVarId) (f : Expr → MetaM Expr) : MetaM MVarId := mvarId.withContext do
  mvarId.checkNotAssigned `grind
  let tag ← mvarId.getTag
  let type ← mvarId.getType
  let typeNew ← f type
  let mvarNew ← mkFreshExprSyntheticOpaqueMVar typeNew tag
  mvarId.assign mvarNew
  return mvarNew.mvarId!

/--
Unfold all `reducible` declarations occurring in `e`.
-/
def unfoldReducible (e : Expr) : MetaM Expr :=
  let pre (e : Expr) : MetaM TransformStep := do
    let .const declName _ := e.getAppFn | return .continue
    unless (← isReducible declName) do return .continue
    let some v ← unfoldDefinition? e | return .continue
    return .visit v
  Core.transform e (pre := pre)

/--
Unfold all `reducible` declarations occurring in the goal's target.
-/
def _root_.Lean.MVarId.unfoldReducible (mvarId : MVarId) : MetaM MVarId :=
  mvarId.transformTarget Grind.unfoldReducible

/--
Abstract nested proofs occurring in the goal's target.
-/
def _root_.Lean.MVarId.abstractNestedProofs (mvarId : MVarId) (mainDeclName : Name) : MetaM MVarId :=
  mvarId.transformTarget (Lean.Meta.abstractNestedProofs mainDeclName)

end Lean.Meta.Grind
