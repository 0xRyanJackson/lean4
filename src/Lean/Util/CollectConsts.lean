/-
Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Lean.Expr
import Lean.Declaration

namespace Lean.CollectConsts

structure State where
  visitedExprPtr  : HashSet USize := {}
  constSet        : HashSet Name := {}
  consts          : Array Name := #[]
  deriving Inhabited

def State.add (s : State) (declName : Name) : State :=
  if s.constSet.contains declName then
    s
  else
    { s with constSet := s.constSet.insert declName, consts := s.consts.push declName }

abbrev Visitor := State → State

mutual
  unsafe def visit (e : Expr) : Visitor := fun s =>
    let addr := ptrAddrUnsafe e
    if s.visitedExprPtr.contains addr then s
    else main e { s with visitedExprPtr := s.visitedExprPtr.insert addr }

  unsafe def main : Expr → Visitor
    | .proj _ _ e      => visit e
    | .forallE _ d b _ => visit b ∘ visit d
    | .lam _ d b _     => visit b ∘ visit d
    | .letE _ t v b _  => visit b ∘ visit v ∘ visit t
    | .app f a         => visit a ∘ visit f
    | .mdata _ b       => visit b
    | .const n _       => fun s => s.add n
    | _                => id
end

end CollectConsts

def collectConsts (s : CollectConsts.State) (e : Expr) : CollectConsts.State :=
  unsafe CollectConsts.main e s

def Declaration.collectConsts (d : Declaration) : Array Name :=
  let s := Id.run <| d.foldExprM Lean.collectConsts {}
  s.consts

end Lean
