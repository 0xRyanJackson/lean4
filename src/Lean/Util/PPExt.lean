/-
Copyright (c) 2020 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
prelude
import Lean.Environment
import Lean.MetavarContext
import Lean.Data.OpenDecl
import Lean.Elab.InfoTree.Types

namespace Lean
register_builtin_option pp.raw : Bool := {
  defValue := false
  group    := "pp"
  descr    := "(pretty printer) print raw expression/syntax tree"
}
register_builtin_option pp.raw.showInfo : Bool := {
  defValue := false
  group    := "pp"
  descr    := "(pretty printer) print `SourceInfo` metadata with raw printer"
}
register_builtin_option pp.raw.maxDepth : Nat := {
  defValue := 32
  group    := "pp"
  descr    := "(pretty printer) maximum `Syntax` depth for raw printer"
}
register_builtin_option pp.rawOnError : Bool := {
  defValue := false
  group    := "pp"
  descr    := "(pretty printer) fallback to 'raw' printer when pretty printer fails"
}

structure PPContext where
  env           : Environment
  mctx          : MetavarContext := {}
  lctx          : LocalContext := {}
  opts          : Options := {}
  currNamespace : Name := Name.anonymous
  openDecls     : List OpenDecl := []

abbrev PrettyPrinter.InfoPerPos := RBMap Nat Elab.Info compare
structure FormatWithInfos where
  fmt : Format
  infos : PrettyPrinter.InfoPerPos
deriving Inhabited

instance : Coe Format FormatWithInfos where
  coe fmt := { fmt, infos := ∅ }

/-- Adds `offset` to each tag in the `Format` -/
-- We could make this a constructor so that repated shifting is more efficient
def _root_.Std.Format.shiftTag (offset : Nat) : Format → Format
  | .tag tag fmt => .tag (tag + offset) (shiftTag offset fmt)
  | .nest i fmt  => .nest i (shiftTag offset fmt)
  | .group fmt b => .group (shiftTag offset fmt) b
  | .append x y  => .append (shiftTag offset x) (shiftTag offset y)
  | fmt => fmt

def FormatWithInfos.append : FormatWithInfos → FormatWithInfos → FormatWithInfos
  | ⟨fmt1, infos1⟩, ⟨fmt2, infos2⟩ => Id.run do
    if infos1.isEmpty then return ⟨fmt1 ++ fmt2, infos2⟩
    if infos2.isEmpty then return ⟨fmt1 ++ fmt2, infos1⟩
    let (offset, _) := infos1.max!
    let fmt2' := fmt2.shiftTag offset
    let infos2' := infos2.mapKeysMono (· + offset)
      (by intros; simp [compare, compareOfLessAndEq]
          all_goals repeat (split <;> ((try rfl); try omega)))
    ⟨fmt1 ++ fmt2', .mergeBy (fun _ _ _ => panic! "disjoint maps overlap") infos1 infos2'⟩

instance AppendFormatWithInfos : Append FormatWithInfos where
  append := FormatWithInfos.append

def FormatWithInfos.nestD : FormatWithInfos → FormatWithInfos
  | ⟨fmt, infos⟩ => ⟨fmt.nestD, infos⟩

def FormatWithInfos.indentD : FormatWithInfos → FormatWithInfos
  | ⟨fmt, infos⟩ => ⟨fmt.indentD, infos⟩

structure PPFns where
  ppExprWithInfos : PPContext → Expr → IO FormatWithInfos
  ppTerm : PPContext → Term → IO Format
  ppGoal : PPContext → MVarId → IO Format
  deriving Inhabited

builtin_initialize ppFnsRef : IO.Ref PPFns ←
  IO.mkRef {
    ppExprWithInfos := fun _ e => return format (toString e)
    ppTerm := fun ctx stx => return stx.raw.formatStx (some <| pp.raw.maxDepth.get ctx.opts)
    ppGoal := fun _ _ => return "goal"
  }

builtin_initialize ppExt : EnvExtension PPFns ←
  registerEnvExtension ppFnsRef.get

def ppExprWithInfos (ctx : PPContext) (e : Expr) : IO FormatWithInfos := do
  let e := instantiateMVarsCore ctx.mctx e |>.1
  if pp.raw.get ctx.opts then
    return format (toString e)
  else
    try
      ppExt.getState ctx.env |>.ppExprWithInfos ctx e
    catch ex =>
      if pp.rawOnError.get ctx.opts then
        pure f!"[Error pretty printing expression: {ex}. Falling back to raw printer.]{Format.line}{e}"
      else
        pure f!"failed to pretty print expression (use 'set_option pp.rawOnError true' for raw representation)"

def ppTerm (ctx : PPContext) (stx : Term) : IO Format :=
  let fmtRaw := fun () => stx.raw.formatStx (some <| pp.raw.maxDepth.get ctx.opts) (pp.raw.showInfo.get ctx.opts)
  if pp.raw.get ctx.opts then
    return fmtRaw ()
  else
    try
      ppExt.getState ctx.env |>.ppTerm ctx stx
    catch ex =>
      if pp.rawOnError.get ctx.opts then
        pure f!"[Error pretty printing syntax: {ex}. Falling back to raw printer.]{Format.line}{fmtRaw ()}"
      else
        pure f!"failed to pretty print term (use 'set_option pp.rawOnError true' for raw representation)"

def ppGoal (ctx : PPContext) (mvarId : MVarId) : IO Format :=
    ppExt.getState ctx.env |>.ppGoal ctx mvarId

end Lean
