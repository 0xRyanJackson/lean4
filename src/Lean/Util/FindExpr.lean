/-
Copyright (c) 2020 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Lean.Expr
import Lean.Util.PtrSet

namespace Lean
namespace Expr
namespace FindImpl

unsafe abbrev FindM (m) := StateT (PtrSet Expr) m

@[inline] unsafe def checkVisited [Monad m] (e : Expr) : OptionT (FindM m) Unit := do
  if (← get).contains e then
    failure
  modify fun s => s.insert e

unsafe def findM? [Monad m] (p : Expr → m Bool) (e : Expr) : OptionT (FindM m) Expr :=
  let rec visit (e : Expr) := do
    checkVisited e
    if ← p e then
      pure e
    else match e with
      | .forallE _ d b _ => visit d <|> visit b
      | .lam _ d b _     => visit d <|> visit b
      | .mdata _ b       => visit b
      | .letE _ t v b _  => visit t <|> visit v <|> visit b
      | .app f a         => visit f <|> visit a
      | .proj _ _ b      => visit b
      | _                => failure
  visit e

unsafe def findUnsafeM? {m} [Monad m] (p : Expr → m Bool) (e : Expr) : m (Option Expr) :=
  findM? p e |>.run' mkPtrSet

@[inline] unsafe def findUnsafe? (p : Expr → Bool) (e : Expr) : Option Expr := findUnsafeM? (m := Id) p e

end FindImpl

/--
Find a subexpression on which the (monadic) predicate `p` returns true.

In compiled code, this is replaced by a fast implementation which uses a cache.

Note that if `p` is not pure (e.g. it uses a cache), then this reference implementation may not
match the compiled behavior! In particular, the compiled implementation will not re-evaluate `p`
on repeated subterms.

Please use this function cautiously; if it becomes a source of bugs, we may remove it.

(See `find?` below for a pure version.)
-/

@[implemented_by FindImpl.findUnsafeM?]
def findM? [Monad m] (p : Expr → m Bool) (e : Expr) : m (Option Expr) := do
  if ← p e then
    return some e
  else match e with
    | .forallE _ d b _ => findM? p d <||> findM? p b
    | .lam _ d b _     => findM? p d <||> findM? p b
    | .mdata _ b       => findM? p b
    | .letE _ t v b _  => findM? p t <||> findM? p v <||> findM? p b
    | .app f a         => findM? p f <||> findM? p a
    | .proj _ _ b      => findM? p b
    | _                => pure none

@[implemented_by FindImpl.findUnsafe?]
def find? (p : Expr → Bool) (e : Expr) : Option Expr := findM? (m := Id) p e

/-- Return true if `e` occurs in `t` -/
def occurs (e : Expr) (t : Expr) : Bool :=
  (t.find? fun s => s == e).isSome


/--
  Return type for `findExt?` function argument.
-/
inductive FindStep where
  /-- Found desired subterm -/ | found
  /-- Search subterms -/ | visit
  /-- Do not search subterms -/ | done

namespace FindExtImpl

unsafe def findM? (p : Expr → FindStep) (e : Expr) : OptionT (FindImpl.FindM Id) Expr :=
  visit e
where
  visitApp (e : Expr) :=
    match e with
    | .app f a .. => visitApp f <|> visit a
    | e => visit e

  visit (e : Expr) := do
    FindImpl.checkVisited e
    match p e with
      | .done  => failure
      | .found => pure e
      | .visit =>
        match e with
        | .forallE _ d b _ => visit d <|> visit b
        | .lam _ d b _     => visit d <|> visit b
        | .mdata _ b       => visit b
        | .letE _ t v b _  => visit t <|> visit v <|> visit b
        | .app ..          => visitApp e
        | .proj _ _ b      => visit b
        | _                => failure

unsafe def findUnsafe? (p : Expr → FindStep) (e : Expr) : Option Expr :=
  Id.run <| findM? p e |>.run' mkPtrSet

end FindExtImpl

/--
  Similar to `find?`, but `p` can return `FindStep.done` to interrupt the search on subterms.
  Remark: Differently from `find?`, we do not invoke `p` for partial applications of an application. -/
@[implemented_by FindExtImpl.findUnsafe?]
opaque findExt? (p : Expr → FindStep) (e : Expr) : Option Expr

end Expr
end Lean
