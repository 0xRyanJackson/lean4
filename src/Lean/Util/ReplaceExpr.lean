/-
Copyright (c) 2020 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura, Gabriel Ebner, Sebastian Ullrich
-/
prelude
import Lean.Expr
import Lean.Util.PtrSet

namespace Lean
namespace Expr

namespace ReplaceImpl

unsafe abbrev ReplaceM := StateM (PtrMap Expr Expr)

unsafe def cache (key : Expr) (exclusive : Bool)  (result : Expr) : ReplaceM Expr := do
  unless exclusive do
    modify (·.insert key result)
  pure result

@[specialize]
unsafe def replaceUnsafeM (f? : Expr → Option Expr) (e : Expr) : ReplaceM Expr := do
  let rec @[specialize] visit (e : Expr) := do
    /-
    TODO: We need better control over RC operations to ensure
    the following (unsafe) optimization is correctly applied.
    Optimization goal: only cache results for shared objects.

    The main problem is that the current code generator ignores borrow annotations
    for code written in Lean. These annotations are only taken into account for extern functions.

    Moveover, the borrow inference heuristic currently tags `e` as "owned" since it may be stored
    in the cache and is used in "update" functions.
    Thus, when visiting `e` sub-expressions the code generator increases their RC
    because we are recursively invoking `visit` :(

    Thus, to fix this issue, we must
    1- Take borrow annotations into account for code written in Lean.
    2- Mark `e` is borrowed (i.e., `(e : @& Expr)`)
    -/
    let excl := isExclusiveUnsafe e
    unless excl do
      if let some result := (← get).find? e then
        return result
    match f? e with
      | some eNew => cache e excl eNew
      | none      => match e with
        | .forallE _ d b _   => cache e excl <| e.updateForallE! (← visit d) (← visit b)
        | .lam _ d b _       => cache e excl <| e.updateLambdaE! (← visit d) (← visit b)
        | .mdata _ b         => cache e excl <| e.updateMData! (← visit b)
        | .letE _ t v b _    => cache e excl <| e.updateLet! (← visit t) (← visit v) (← visit b)
        | .app f a           => cache e excl <| e.updateApp! (← visit f) (← visit a)
        | .proj _ _ b        => cache e excl <| e.updateProj! (← visit b)
        | e                  => return e
  visit e

@[inline]
unsafe def replaceUnsafe (f? : Expr → Option Expr) (e : Expr) : Expr :=
  (replaceUnsafeM f? e).run' mkPtrMap

end ReplaceImpl

/- TODO: use withPtrAddr, withPtrEq to avoid unsafe tricks above.
   We also need an invariant at `State` and proofs for the `uget` operations. -/

@[specialize]
def replaceNoCache (f? : Expr → Option Expr) (e : Expr) : Expr :=
  match f? e with
  | some eNew => eNew
  | none      => match e with
    | .forallE _ d b _ => let d := replaceNoCache f? d; let b := replaceNoCache f? b; e.updateForallE! d b
    | .lam _ d b _     => let d := replaceNoCache f? d; let b := replaceNoCache f? b; e.updateLambdaE! d b
    | .mdata _ b       => let b := replaceNoCache f? b; e.updateMData! b
    | .letE _ t v b _  => let t := replaceNoCache f? t; let v := replaceNoCache f? v; let b := replaceNoCache f? b; e.updateLet! t v b
    | .app f a         => let f := replaceNoCache f? f; let a := replaceNoCache f? a; e.updateApp! f a
    | .proj _ _ b      => let b := replaceNoCache f? b; e.updateProj! b
    | e                => e


@[extern "lean_replace_expr"]
opaque replaceImpl (f? : @& (Expr → Option Expr)) (e : @& Expr) : Expr

@[implemented_by ReplaceImpl.replaceUnsafe]
def replace (f? : Expr → Option Expr) (e : Expr) : Expr :=
  e.replaceNoCache f?
