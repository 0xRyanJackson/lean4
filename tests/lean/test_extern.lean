import Lean.Util.TestExtern

instance : BEq ByteArray where
  beq x y := x.data == y.data

test_extern Nat.add 12 37
test_extern 4 + 5

test_extern ByteArray.copySlice ⟨#[1,2,3]⟩ 1 ⟨#[4, 5, 6, 7, 8, 9, 10, 11, 12, 13]⟩ 0 6

@[extern "lean_nat_add"]
protected def Nat.not_add : (@& Nat) → (@& Nat) → Nat
  | a, Nat.zero   => a
  | a, Nat.succ b => Nat.succ (Nat.add a b)

test_extern Nat.not_add 0 0
