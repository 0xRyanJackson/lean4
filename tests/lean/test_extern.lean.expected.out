test_extern.lean:6:0-6:17: error: test_extern: HAdd.hAdd does not have an @[extern] attribute or @[implemented_by] attribute
test_extern.lean:8:0-8:86: error: native implementation did not agree with reference implementation!
Compare the outputs of:
#eval ByteArray.copySlice { data := #[1, 2, 3] } 1 { data := #[4, 5, 6, 7, 8, 9, 10, 11, 12, 13] } 0 6
 and
#eval {
  data :=
    Array.extract { data := #[4, 5, 6, 7, 8, 9, 10, 11, 12, 13] }.data 0 0 ++
        Array.extract { data := #[1, 2, 3] }.data 1 (1 + 6) ++
      Array.extract { data := #[4, 5, 6, 7, 8, 9, 10, 11, 12, 13] }.data (0 + 6)
        (Array.size { data := #[4, 5, 6, 7, 8, 9, 10, 11, 12, 13] }.data) }
test_extern.lean:15:0-15:13: error: native implementation did not agree with reference implementation!
Compare the outputs of: #eval g  and #eval 4
